ninja-core/src/main/java/ninja/ReverseRouter.java

```diff
@@ -0,0 +1,421 @@
+        
+        private Builder setPath(String name, Object value, boolean raw) {
+            Objects.requireNonNull(name, "name required");
+            Objects.requireNonNull(value, "value required");
+
+            if (route.getParameters() == null || !route.getParameters().containsKey(name)) {
+                throw new IllegalArgumentException("Reverse route " + route.getUri()
+                    + " does not have a path parameter '" + name + "'");
+            }
+            
+            if (this.paths == null) {
+                this.paths = new HashMap<>();
+            }
+            
+            this.paths.put(name, safeValue(value, raw));
+            
+            return this;
+        }
+        
+        /**
+         * Add a parameter as a query string value. This method will URL encode
+         * the values when building the final result.
+         * @param name The query string parameter name
+         * @param value The query string parameter value
+         * @return A reference to this builder
+         * @see #rawQuery(java.lang.String, java.lang.Object) 
+         */
+        public Builder query(String name, Object value) {
+            return setQuery(name, value, false);
+        }
+        
+        /**
+         * Identical to <code>query</code> except the query string value will
+         * NOT be url encoded when building the final url.
+         * @param name The query string parameter name
+         * @param value The query string parameter value
+         * @return A reference to this builder
+         * @see #query(java.lang.String, java.lang.Object) 
+         */
+        public Builder rawQuery(String name, Object value) {
+            return setQuery(name, value, true);
+        }
+        
+        private Builder setQuery(String name, Object value, boolean raw) {
+            Objects.requireNonNull(name, "name required");
+
+            if (this.queries == null) {
+                // retain ordering
+                this.queries = new LinkedHashMap<>();
+            }
+            
+            this.queries.put(name, safeValue(value, raw));
+            
+            return this;
+        }
+        
+        /**
+         * Add a map containing pairs with replacements for either path or query
+         * string parameters. This method will URL encode the values when building
+         * the final result.  When replacing path parameters it's recommended
+         * to use the <code>path</code> method since it will validate the parameter
+         * exists, while this method would fallback to adding the value as a query
+         * string parameter if the path parameter did not exist.
+         * @param parameterMap The map containing pairs of name and values.
+         * @return A reference to this builder
+         * @see #path(java.lang.String, java.lang.Object)
+         * @see #query(java.lang.String, java.lang.Object) 
+         */
+        public Builder params(Map<String,Object> parameterMap) {
```

Comment:
We could remove that method  - what do you think? param and query are already perfect...